// @here/harp-vectortile-datasource/lib/OmvPoliticalViewFeatureModifier.js
"use strict";
/*
 * Copyright (C) 2017-2019 HERE Europe B.V.
 * Licensed under Apache 2.0, see full license in LICENSE
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OmvPoliticalViewFeatureModifier = void 0;
const harp_utils_1 = require("@here/harp-utils");
const logger = harp_utils_1.LoggerManager.instance.create("OmvPoliticalViewFeatureModifier");
/**
 * Modifies the MapEnv of the Vector Tiles in Tilezen format with different POV.
 *
 * This feature modifier updates feature properties according to different political
 * point of view.
 * Political views (or alternate point of views) are supported in Tilezen by adding
 * country posix (lower-case ISO 3166-1 alpha-2 compliant) to __default__ property name.
 * For example country borders (__boundaries__ layer) may have both __kind__ property for
 * default (commonly accepted point of view) and __kind:xx__ for alternate points of view.
 * This way disputed borders may be visible or not for certain regions and different
 * users (clients).
 *
 * @hidden
 */
class OmvPoliticalViewFeatureModifier {
    /**
     * C-tor.
     *
     * @param pov - The code of the country (in lower-case ISO 3166-1 alpha-2 format) which
     * point of view should be taken into account.
     */
    constructor(pov) {
        this.m_countryCode = pov;
    }
    /**
     * Simply passes all points to rendering, points features does not support PoliticalView.
     *
     * @param layer - Current layer.
     * @param env - Properties of point feature.
     * @param level - Level of tile.
     * @returns always `true` to pass feature.
     */
    doProcessPointFeature(layer, env, level) {
        return true;
    }
    /**
     * Implements line features processing changing "kind" attribute depending on point of view.
     *
     * Currently only line features support different point of view.
     * @param layer - The name of the layer.
     * @param env - The environment to use.
     * @returns always `true` to pass lines for rendering.
     */
    doProcessLineFeature(layer, env, level) {
        this.rewriteEnvironment(layer, env);
        return true;
    }
    /**
     * Simply pass all polygons to rendering, this feature does not support PoliticalView yet.
     *
     * @param layer - Current layer.
     * @param env - Properties of polygon feature.
     * @param level - Level of tile.
     * @returns `true` to pass feature.
     */
    doProcessPolygonFeature(layer, env, level) {
        return true;
    }
    /**
     * Rewrites the Environment to match the different points of view.
     *
     * @param layer - The layer name.
     * @param env - The environment to use.
     */
    rewriteEnvironment(layer, env) {
        // For now we need to rewrite "boundaries" layer only.
        if (this.isPoliticalViewLayer(layer)) {
            this.updateEnvironment(env, this.m_countryCode, "kind");
        }
    }
    updateEnvironment(env, countryCode, propName) {
        const value = this.getAlternativePov(env, countryCode, propName);
        if (value !== undefined) {
            env.entries[propName] = value;
        }
    }
    getAlternativePov(env, countryCode, propName) {
        logger.log("Get alternate POV: ", JSON.stringify(env));
        const cc = countryCode;
        const value = env.lookup(`${propName}:${cc}`);
        logger.log("Lookup POV: ", `${propName}:${cc}`, value);
        if (typeof value === "string" && value.length > 0) {
            logger.log("Found POV: ", `${propName}:${cc}`, value);
            return value;
        }
        else {
            return undefined;
        }
    }
    isPoliticalViewLayer(layer) {
        return layer === "boundaries";
    }
}
exports.OmvPoliticalViewFeatureModifier = OmvPoliticalViewFeatureModifier;
